Public/Src/Cache/ContentStore/Distributed/NuCache/InMemory/MemoryContentLocationDatabase.cs

```diff
@@ -49,6 +51,19 @@ internal override void Persist(OperationContext context, ShortHash hash, Content
             _map.AddOrUpdate(hash, key => entry, (key, old) => entry);
         }
 
+        /// <inheritdoc />

```

Comment:
Just a generic comment on all the cache interfaces.

I have started seeing a pattern of having a base function and than have one derived class implement them where the others return null.

Does this warrant a capability based design,  or perhaps more specific interfaces?

Reply:
In this particular case, the returning of null is ReSharper filling with null, why doesn't matter because this is a mock class and we actually already have a class that serves an IMemoizationSession from memory elsewhere. I'd check if that's not the case as well for the other interfaces you mentioned.

Regarding your question, I'm not sure what you meant by capability-based design. I'd argue against more interfaces any day, as I feel we have enough of them already, many of which make things more confusing and restraining rather than allow for more concise expression of ideas.

See for example IKeyValueStore, which is only used to represent a RocksDB store; it turns into a constraint: many things that RocksDB supports are not usable because you'd need to create a "generic" interface to those, and you also have to provide string and byte[] methods for every single function call, which is kind of pointless since it just moves the type conversion to the class.

---
In reply to: [302163716](https://github.com/microsoft/BuildXL/pull/571#discussion_r302163716) [](ancestors = 302163716)